pageserver/src/layered_repository/layer_map.rs

```diff
@@ -102,34 +101,35 @@ impl LayerMap {
         NUM_INMEMORY_LAYERS.inc();
     }
 
-    /// Remove the oldest in-memory layer
-    pub fn pop_oldest_open(&mut self) {
-        // Pop it from the binary heap
-        let oldest_entry = self.open_layers.pop().unwrap();
+    /// Remove an open in-memory layer
+    pub fn remove_open(&mut self, layer_id: LayerId) {
+        // Note: we don't try to remove the entry from the binary heap.
+        // It will be removed lazily by peek_oldest_open() when it's made it to
+        // the top of the heap.
 
         let layer_opt = {
             let mut global_map = GLOBAL_LAYER_MAP.write().unwrap();
-            let layer_opt = global_map.get(&oldest_entry.layer_id);
-            global_map.remove(&oldest_entry.layer_id);
-            // TODO it's bad that a ref can still exist after being evicted from global map
+            let layer_opt = global_map.get(&layer_id);
+            global_map.remove(&layer_id);
+            // TODO it's bad that a ref can still exist after being evicted from cache
             layer_opt
         };
 
         if let Some(layer) = layer_opt {
             let mut segentry = self.segs.get_mut(&layer.get_seg_tag()).unwrap();
 
-            // Also remove it from the SegEntry of this segment
-            if segentry.open_layer_id.unwrap() == oldest_entry.layer_id {
+            if segentry.open_layer_id == Some(layer_id) {
+                // Also remove it from the SegEntry of this segment
                 segentry.open_layer_id = None;
             } else {
                 // We could have already updated segentry.open for
                 // dropped (non-writeable) layer. This is fine.
                 assert!(!layer.is_writeable());
                 assert!(layer.is_dropped());
```

Comment:
I was thinking about this a bit more. Is it possible that the older layer could be evicted from the global map, causing !is_writeable, but also !is_dropped ? @hlinnaka